JOSS Submission (#503)
pangeo-data/climpred: joss/paper.md

---
title: 'climpred: Verification of weather and climate forecasts'
tags:
  - python
  - climate
  - forecasting
  - geospatial
  - big data
authors:
  - name: Riley X. Brady
    orcid: 0000-0002-2309-8245
    affiliation: 1
  - name: Aaron Spring
    orcid: 0000-0003-0216-2241

>     affiliation: 2

Reviewer: ```suggestion
    affiliation: 2, 3
```

affiliations:
 - name: Department of Atmospheric and Oceanic Sciences and Institute of Arctic and Alpine Research, University of Colorado Boulder, Boulder, Colorado USA
   index: 1

>  - name: International Max Planck Research School on Earth System Modelling and Max Planck Institute for Meteorology, Hamburg, Germany
>    index: 2
> date: 16 October 2020

Reviewer: ```suggestion
 - name: Max Planck Institute for Meteorology, Hamburg, Germany
   index: 2
 - name: International Max Planck Research School on Earth System Modelling, Hamburg, Germany
   index: 3
date: 16 October 2020
```

bibliography: paper.bib
---

<!-- 982 words total -->

# Summary
<!-- 71 words -->
Predicting extreme events and variations in weather and climate provides crucial
information for economic, social, and environmental decision-making. However,
quantifying prediction skill for multi-dimensional geospatial model output is
computationally expensive and a difficult coding challenge. The large datasets
(order gigabytes to terabytes) require parallel and out-of-memory computing to be
analyzed efficiently. Further, aligning the many forecast initializations with differing
observational products is a straight-forward, but exhausting and error-prone exercise
for researchers.

<!-- 68 words -->
To simplify and standardize forecast verification across scales from hourly weather to
decadal climate forecasts, we built `climpred`: a community-driven python package for
computationally efficient and methodologically consistent verification of ensemble
prediction models. The code base is maintained through open-source development. It
leverages `xarray` [@Hoyer:2017] to anticipate core prediction ensemble dimensions

> (ensemble `member`, `init`ialization date and `lead` time) and `dask` [@dask] to

Reviewer: we could also link to https://conference.scipy.org/proceedings/scipy2015/matthew_rocklin.html

Reviewer: I dont like the `team` rendering

Author: This is the suggested citation from the `dask` website. We can just do both.

> (ensemble `member`, `init`ialization date and `lead` time) and `dask` [@dask] to

Reviewer: ```suggestion
(ensemble `member`, `init`ialization date and `lead` time) and `dask` [@Rocklin:2015] to
```

Reviewer: or both

 <!-- 54 words -->
`climpred` aims to offer a comprehensive set of analysis tools for assessing the quality
of dynamical forecasts relative to verification products (_e.g._, observations,
reanalysis products, control simulations). The package includes a suite of deterministic
and probabilistic verification metrics that are constantly expanded by the community and
are generally organized in our companion package, `xskillscore`.

<!-- 161 words -->
# Statement of Need
While other climate verification packages exist (_e.g._, `s2dverification`
[@Manubens:2018] written in R and `MurCSS` [@Illing:2014] written with python-based
`CDO`-bindings [@CDO]), `climpred` is unique for many reasons. (1) It spans broad
temporal scales of prediction, supporting the weather, subseasonal-to-seasonal (S2S),
and seasonal-to-decadal (S2D) communities. (2) `climpred` supports `dask` [@dask] and
thus works across all computational scales, from personal laptops to supercomputers
(HPC), and in the cloud. This leads to, for example, verification of a global 5°x5°
resolution climate prediction in less than one second, compared to the 8 minutes

> required in `MurCSS`.  (3) `climpred` is highly modular and supports the research

Author: Do you think it's okay to put it this way? Don't want this to be offensive to the MurCSS folks for all the work they did, by singling them out.

Author: Ryan Abernathy suggests it's okay, since that's what pushes software forward. The way I wrote it in the medium article is this, which nods to their contributions and the bias in benchmarking. Maybe we can adapt that shortly here:

> Due to the use of vectorized numpy code and parallel, out-of-memory computing with dask , climpred verifies forecasts more quickly than other forecast verification packages. This enables a truly interactive analysis experience for users. For example, climpred can verify a 5° x 5° global forecast model in 0.5 seconds, relative to the 8 minutes required by MurCSS. This means that a user can quickly change verification metrics and rapidly assess their results. However, note that [benchmarking is inherently biased](https://matthewrocklin.com/blog/work/2017/03/09/biased-benchmarks) and MurCSS is quite valuable for their rigorous replication of decadal climate prediction metrics and their seamless integration with the MiKlip decadal prediction project at the Max Planck Institute for Meteorology.

Reviewer: `offensive to the MurCSS folks for all the work` thats also my concern. but also Ryans comment is true. however, I would say something like 5 or a few seconds. the 8 minutes verify ACC, conditional and unconditional bias and CRPSS_ES over three leads. my CPRSS_ES examples over 10 leads took 10 seconds. so overall I think ~5 seconds is a fair comparison estimate from our side. I wouldnt go into details on this but just say `a few seconds`.

dynamics of the real world. Skill quantification is accomplished by considering one
ensemble member as the verification data and the remaining members as the forecasts
[@Griffies:1997].

<!-- 362 words -->
# Climpred Classes and Object-Oriented Verification
`climpred` supports both prediction system formats, offering `HindcastEnsemble` and
`PerfectModelEnsemble` objects. `HindcastEnsemble` is instantiated with an `initialized`
hindcast ensemble dataset and requires an `observation`al dataset against which to
verify. `PerfectModelEnsemble` is instantiated with an `initialized` perfect-model
ensemble dataset and also accepts a `control` dataset against which to evaluate
forecasts. Both objects can also track an `uninitialized` dataset, which represents a
historical simulation that evolves solely due to random internal climate variability or
can be used to isolate the influence of external forcing [e.g., @Kay:2014].

Assessing skill for `PredictionEnsemble` objects (the parent class to `HindcastEnsemble`
and `PerfectModelEnsemble`) is standardized into a one-liner:

```python
PredictionEnsemble.verify(

>     # Score forecast using the Anomaly Correlation Coefficient.

Author: Needs to be above the keyword arg or it runs off the page when rendered.

---
title: 'climpred: Verification of weather and climate forecasts'
tags:
  - python
  - climate
  - forecasting
  - geospatial
  - big data
authors:
  - name: Riley X. Brady
    orcid: 0000-0002-2309-8245
    affiliation: 1
  - name: Aaron Spring
    orcid: 0000-0003-0216-2241
    affiliation: 2, 3
affiliations:
 - name: Department of Atmospheric and Oceanic Sciences and Institute of Arctic and Alpine Research, University of Colorado Boulder, Boulder, Colorado USA
   index: 1
 - name: Max Planck Institute for Meteorology, Hamburg, Germany
   index: 2
 - name: International Max Planck Research School on Earth System Modelling, Hamburg, Germany
   index: 3
date: 16 October 2020
bibliography: paper.bib
---

<!-- 982 words total -->

# Summary
<!-- 71 words -->
Predicting extreme events and variations in weather and climate provides crucial

> information for economic, social, and environmental decision-making. However,

Reviewer: ```suggestion
information for economic, social, and environmental decision-making [@Merryfield:2020]. However,
```

reanalysis products, control simulations). The package includes a suite of deterministic
and probabilistic verification metrics that are constantly expanded by the community and
are generally organized in our companion package, `xskillscore`.

<!-- 161 words -->
# Statement of Need
While other climate verification packages exist (_e.g._, `s2dverification`
[@Manubens:2018] written in R and `MurCSS` [@Illing:2014] written with python-based
`CDO`-bindings [@CDO]), `climpred` is unique for many reasons. (1) It spans broad
temporal scales of prediction, supporting the weather, subseasonal-to-seasonal (S2S),
and seasonal-to-decadal (S2D) communities. (2) `climpred` supports `dask` [@dask] and
thus works across all computational scales, from personal laptops to supercomputers
(HPC), and in the cloud. This leads to, for example, verification of a global 5°x5°
resolution climate prediction in less than one second, compared to the 8 minutes
required in `MurCSS`.  (3) `climpred` is highly modular and supports the research
process from end-to-end, from loading in model output, to interactive pre-processing and
analysis, to visualization. (4) `climpred` is part of and benefits from the wider

> scientific python community, `pangeo` [@Eynard:2019]. A wide adoption of `climpred`

Reviewer: is this the best pangeo reference there is?

Reviewer: I only know of this one as an alternative: 
Odaka, T. E., Banihirwe, A., Eynard-Bontemps, G., Ponte, A., Maze, G., Paul, K., Baker, J., & Abernathey, R. (2020). The Pangeo Ecosystem: Interactive Computing Tools for the Geosciences: Benchmarking on HPC. G. Juckeland & S. Chandrasekaran (Eds.), Tools and Techniques for High Performance Computing (pp. 190–204). Cham: Springer International Publishing. doi: 10/ggqs7s

Author: I spoke with Ryan on gitter and that's what he suggested. Said they're still working on a formal paper.

could standardize prediction model evaluation and make verification reproducible
[@Irving:2015]. (5) The `climpred` documentation serves as a repository of unified
analysis methods through `jupyter` notebook [@Kluyver:2016] examples and collects
relevant references and literature.

<!-- 207 words -->
# Prediction Simulation Types

> Weather and climate modeling institutions typically run so-called “hindcasts," where

Reviewer: ```suggestion
Weather and climate modeling institutions typically run so-called “hindcasts", where
```

> Weather and climate modeling institutions typically run so-called “hindcasts", where
> they retrospectively initialize dynamical models from many past observed climate states
> [@Meehl:2009]. Initializations are then slightly perturbed to generate an ensemble of
> forecasts that diverge solely due to their sensitive dependence on initial conditions
> [@Lorenz:1963]. Hindcasts are evaluated by using some statistical metric to score their

Reviewer: ```suggestion
Weather and climate modeling institutions typically run so-called “hindcasts", where
dynamical models are retrospectively initialized from many past observed climate states
[@Meehl:2009]. Initializations are then slightly perturbed to generate an ensemble of
forecasts that diverge solely due to their sensitive dependence on initial conditions
[@Lorenz:1963]. Hindcasts are evaluated by using some statistical metric to score their
```

with observational dates in the academic literature. The authors of `climpred` thus
identified three techniques, which can be selected by the user: (1) Maximize the degrees
of freedom by selecting all initialization dates that verify with the available
observations at each lead. In turn, initializations and verification dates
are not held constant for each lead. (2) Use the identical set of initializations that
can verify over the given observational window at all leads. However, the verification
dates change at each lead. (3) Use the identical verification window at each lead, while
allowing the set of initializations used at each lead to change. These strategies are
shown graphically and explained in more detail in the documentation. Note that
`climpred` offers extensive analysis functionality in addition to forecast verification,
such as spatiotemporal smoothing [@Goddard:2013], bias removal [@Boer:2016],
significance testing [@Goddard:2013; @Boer:2016; @DelSole:2016], and a graphics library.

<!-- 59 words -->
# Use in Academic Literature
`climpred` has been used to drive analysis in three academic papers, to the author’s

> knowledge. @Brady:2020 used the hindcast module to highlight multi-year predictability

Reviewer: ```suggestion
knowledge. @Brady:2020 used the `HindcastEnsemble` to highlight multi-year predictability
```

> `climpred` has been used to drive analysis in three academic papers, to the author’s
> knowledge. @Brady:2020 used the hindcast module to highlight multi-year predictability

Reviewer: ```suggestion
`climpred` has been used to drive analysis in three academic papers so far. @Brady:2020 used the hindcast module to highlight multi-year predictability
```

> knowledge. @Brady:2020 used the hindcast module to highlight multi-year predictability
> of ocean acidification in the California Current; @Spring:2020 used the perfect-model
> module to highlight predictability horizons in the global carbon cycle; and
> @Krumhardt:2020 used the hindcast module illuminate multi-year predictability in marine

Reviewer: ```suggestion
knowledge. @Brady:2020 used `HindcastEnsemble` to highlight multi-year predictability
of ocean acidification in the California Current; @Spring:2020 used `PerfectModelEnsemble` to highlight predictability horizons in the global carbon cycle; and
@Krumhardt:2020 used `HindcastEnsemble` to illuminate multi-year predictability in marine
```
